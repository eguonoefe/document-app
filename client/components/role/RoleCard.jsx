import React from 'react';
import PropTypes from 'prop-types';
import { Modal } from 'react-materialize';
import InputField from '../common/InputField';
import parseDate from '../../utilities/parseDate';

/**
* Role Card for viewing a single user
*
* @param {Object} props { title, id, description,
*  updatedAt, onChange, onSubmit, deleteRole }
* @returns {String} HTML markup for the Role Card
*/
const RoleCard = ({
    title, id, description, updatedAt, onChange, onSubmit, deleteRole }) => (
    <div>
      <h5>{title}</h5>
      <p>{description}</p>
      <p>Last Edited: {parseDate(updatedAt)}</p>
      <p />
      <p />
      <div className="col l6 m6 s12 left">
        <Modal
          trigger={
            <a
              className="waves-effect btn button-design"
              data-target="passwordModal" id="delete-role">
              Delete
                </a>}
          actions={
            <div>
              <button
                className="btn btn-flat modal-action modal-close"
                name={id} onClick={deleteRole}
                id="delete">
                Delete
                </button>
              <button
                className="btn btn-flat modal-action modal-close left">
                Cancel
                  </button>
            </div>}>
          <div >
            <h5
              className="center">
              Are you sure you want to delete the user
                </h5>
          </div>
        </Modal>
      </div>
      <div className="col l6 m6 s12 right">
        <Modal
          trigger={
            <a
              className="waves-effect btn button-design"
              data-target="passwordModal"
              id="update-role">
              Update
          </a>
          }
          actions={
            <div>
              <button
                className="btn btn-flat modal-action modal-close"
                type="submit"
                onClick={onSubmit}
                name={id}
                id="update">
                Update
                </button>
              <button
                className="btn btn-flat modal-action modal-close left">
                Cancel
                  </button>
            </div>}>
          <div>
            <h5>Update Role</h5>
            <InputField
              name="title"
              placeholder={title}
              className="validate form-design"
              type="text"
              onChange={onChange} />
            <InputField
              name="description"
              placeholder={description}
              className="validate form-design"
              type="text"
              onChange={onChange} />
          </div>
        </Modal>
      </div>
    </div>
  );

// Set Props
RoleCard.propTypes = {
  title: PropTypes.string,
  description: PropTypes.string,
  updatedAt: PropTypes.string,
  onChange: PropTypes.func,
  onSubmit: PropTypes.func,
  deleteRole: PropTypes.func,
  id: PropTypes.number
};

export default RoleCard;
