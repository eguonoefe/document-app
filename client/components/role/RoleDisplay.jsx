import React from 'react';
import PropTypes from 'prop-types';

import InputField from '../common/InputField';

/**
 * Role Display component for creating a role
 *
 * @param {Object} props { onChange, onSubmit, role }
 * @returns {String} HTML markup for the Role Display
 */
const RoleDisplay = ({ onChange, onSubmit, role }) => (
  <div>
    <div>
      <h5>Create Role</h5>
      <InputField
        name="title"
        value={role.title}
        placeholder="role Title"
        className="validate form-design"
        type="text" onChange={onChange} />
      <InputField
        name="description"
        value={role.descripition}
        placeholder="role Content"
        className="validate form-design"
        type="text" onChange={onChange} />
      <button
        className="waves-effect btn button-design"
        type="submit" onClick={onSubmit} id="save-role">
        Save
        </button>
    </div>
  </div>
);

RoleDisplay.propTypes = {
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  role: PropTypes.object.isRequired
};

export default RoleDisplay;
