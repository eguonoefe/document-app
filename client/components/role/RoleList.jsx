import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import RoleActions from '../../actions/RoleActions';
import RoleCard from './RoleCard';

const viewRole = RoleActions.viewRole;
const updateRole = RoleActions.updateRole;
const deleteRole = RoleActions.deleteRole;
/**
 * @class RoleList
 * @extends {Component}
 */
export class RoleList extends Component {
  /**
   * Creates an instance of RoleList.
   * Binds class methods
   * @param {Object} props -
   * @memberof RoleList
   */
  constructor(props) {
    super(props);
    this.state = {
      roles: [],
      role: {}
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.deleteRole = this.deleteRole.bind(this);
  }

  /**
   * Makes action call to retrieve list of roles
   *
   * @return {void}
   * @memberof RoleList
   */
  componentDidMount() {
    this.updateRoleList();
  }
  /**
   * Update the state if the props are changed
   *
   * @return {void}
   * @param {Object} nextProps -
   * @memberof RoleList
   */
  componentWillReceiveProps(nextProps) {
    if (this.props.roleList !== nextProps.roleList) {
      this.setState({
        roles: nextProps.roleList
      });
    }
  }
  /**
  * Sets the event value to the state
  * @return {void}
  * @param {Object} event The event of the HTML component
  * @memberof RoleContainer
  */
  onChange(event) {
    const field = event.target.name;
    this.setState({ [field]: event.target.value });
  }

  /**
   * Makes an action call to get all roles
   * Sets list of role to the state
   *
   * @return {void}
   * @memberof RoleList
   */
  updateRoleList() {
    this.props.viewRole()
      .then(() => {
        if (this.props.role.message) {
          return Materialize.toast(this.props.role.message,
            2000, 'indigo darken-4 white-text rounded');
        }
        this.setState({
          roles: this.props.roleList
        });
      });
  }
  /**
  * Makes an action call to delete a role
  * Toasts the error/success message
  *
  * @return {void}
  * @param {Object} event -
  * @memberof RoleList
  */
  deleteRole(event) {
    this.props.deleteRole(event.target.name).then(() => {
      this.updateRoleList();
      this.props.history.push(`${this.props.match.url}`);
    });
  }

  /**
  * Makes an action call to update a role
  * Toasts error/success message
  * Sets updated role to state
  *
  * @param {Object} event -
  * @return {void}
  * @memberof RoleList
  */
  onSubmit(event) {
    const updatedRole = {
      title: this.state.title,
      description: this.state.description,
    };
    this.props.updateRole(updatedRole, event.target.name)
      .then(() => {
        if (this.props.role.message) {
          return Materialize.toast(this.props.role.message,
            2000, 'indigo darken-4 white-text rounded');
        }
        Materialize.toast('Role successfully updated',
          2000, 'indigo darken-4 white-text rounded');
        this.updateRoleList();
        this.props.history.push(`${this.props.match.url}`);
      });
  }

  /**
   * Renders Rolelist Component
   *
   * @returns {String} HMTL markup for the RoleList
   * @memberof RoleList
   */
  render() {
    return (
      <div className="document-list">
        <div className="container">
          <div className="row">
            <div className="document-list-view">
              <div className="col l6 m6 s12">
                <div className=" card-panel hoverable">
                  <h5>All Roles</h5>
                  <div className="divider" />
                  <div className="scrollable">
                    {this.state.roles.map(Role => (
                      <RoleCard
                        key={Role.id} {...Role}
                        onSubmit={this.onSubmit}
                        onChange={this.onChange}
                        deleteRole={this.deleteRole}
                        match={this.props.match} />
                    ))}
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapPropsToState = state => (
  {
    role: state.role,
    roleList: state.role.roleList,
    access: state.access
  }
);

RoleList.propTypes = {
  viewRole: PropTypes.func,
  updateRole: PropTypes.func,
  deleteRole: PropTypes.func,
  roleList: PropTypes.array,
  role: PropTypes.object,
  match: PropTypes.object,
  history: PropTypes.object
};

export default connect(
  mapPropsToState, { viewRole, updateRole, deleteRole })(withRouter(RoleList));
