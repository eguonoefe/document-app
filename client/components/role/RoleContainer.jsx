import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';

import RoleDisplay from './RoleDisplay';
import RoleActions from '../../actions/RoleActions';
import validate from '../../utilities/validate';

const createRole = RoleActions.createRole;

/**
 * @class RoleContainer
 * @extends {Component}
 */
export class RoleContainer extends Component {
  /**
   * Creates an instance of RoleContainer.
   * Binds the functions to the class
   * @param {Object} props -
   * @memberof RoleContainer
   */
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      description: ''
    };
    this.onSubmit = this.onSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
  }

  /**
  * Validates the RoleDisplay input fields
  * Makes an action call to create a new role
  * Toasts the error/success message
  *
  * @return {void}
  * @memberof RoleContainer
  */
  onSubmit() {
    try {
      const { valid } = validate.validateSaveRole(this.state);
      if (!valid) {
        throw new Error('No field should be left blank');
      }
      this.props.createRole(this.state)
        .then(() => {
          if (this.props.role.message) {
            return Materialize.toast(
              this.props.role.message,
              2000, 'indigo darken-4 white-text rounded');
          }
          Materialize.toast(
            'Role created', 2000, 'indigo darken-4 white-text rounded');
          this.props.history.push('/dashboard/all-roles');
        });
    } catch (err) {
      Materialize.toast(err.message, 3000,
        'indigo darken-4 white-text rounded');
    }
  }
  /**
  * Sets the event value to the state
  * @return {void}
  * @param {Object} event The event of the HTML component
  * @memberof RoleContainer
  */
  onChange(event) {
    const field = event.target.name;
    this.setState({ [field]: event.target.value });
  }


  /**
   * @returns {jsx} -
   * @memberof RoleContainer
   */
  render() {
    return (
      <RoleDisplay
        onSubmit={this.onSubmit}
        onChange={this.onChange}
        role={this.state} />
    );
  }
}

const mapPropsToState = state => (
  {
    role: state.role
  }
);

RoleContainer.propTypes = {
  createRole: PropTypes.func.isRequired,
  role: PropTypes.object,
  history: PropTypes.object
};

export default connect(
  mapPropsToState, { createRole })(withRouter(RoleContainer));
