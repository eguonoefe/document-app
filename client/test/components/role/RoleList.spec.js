/* global expect jest test */
import React from 'react';
import { shallow, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { RoleList } from '../../../components/role/RoleList';

configure({ adapter: new Adapter() });

describe('Role List', () => {
  const props = {
    access: {
      isAuthenticated: true,
      user: {
        id: 1, email: 'efeeguono@gmail.com', roleId: 1
      }
    },
    roleList: [{ title: 'Manny' },
    { title: 'John' }, { title: 'move' }],
    role: { title: 'Eguono', id: 3, description: 'john' },
    viewRole: jest.fn(() => {
      return Promise.resolve();
    }),
    updateRole: jest.fn(() => {
      return Promise.resolve();
    }),
    deleteRole: jest.fn(() => {
      return Promise.resolve();
    }),
    history: {
      push: jest.fn()
    },
    match: {
      url: ''
    },
  };
  const component = shallow(<RoleList {...props} />);
  beforeEach(() => {
    global.Materialize = { toast: jest.fn() };
  });
  describe('Container', () => {
    test('should match snapshot of document list', () => {
      const tree = toJson(component);
      expect(tree).toMatchSnapshot();
    });
    test('onChange function should run when called', () => {
      component.instance().onChange(
        { target: { value: 'eguono', name: 'title' } });
      component.instance().onChange(
        { target: { value: 'efe', name: 'description' } });
      expect(component.state('title')).toEqual('eguono');
      expect(component.state('description')).toEqual('efe');
      const spy = jest.spyOn(component.instance(), 'onSubmit');
      component.instance().onSubmit({
        target: { name: 2 }
      });
      expect(spy).toHaveBeenCalled();
    });
    test('deleteDocument function should run when called', () => {
      const spy = jest.spyOn(component.instance(), 'deleteRole');
      component.instance().deleteRole({ target: { value: 3, name: 'title' } });
      expect(spy).toHaveBeenCalled();
    });
    test('componentWillRecieveProps will run if new props are added', () => {
      const spy = jest.spyOn(component.instance(), 'componentWillReceiveProps');
      const roleList = [{ title: 'john' }];
      component.setProps({ roleList });
      expect(spy).toHaveBeenCalled();
    });
  });
});
