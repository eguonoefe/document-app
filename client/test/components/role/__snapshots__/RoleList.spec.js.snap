// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Role List Container should match snapshot of document list 1`] = `
<div
  className="document-list"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="document-list-view"
      >
        <div
          className="col l6 m6 s12"
        >
          <div
            className=" card-panel hoverable"
          >
            <h5>
              All Roles
            </h5>
            <div
              className="divider"
            />
            <div
              className="scrollable"
            >
              <RoleCard
                deleteRole={[Function]}
                match={
                  Object {
                    "url": "",
                  }
                }
                onChange={[Function]}
                onSubmit={[Function]}
                title="Manny"
              />
              <RoleCard
                deleteRole={[Function]}
                match={
                  Object {
                    "url": "",
                  }
                }
                onChange={[Function]}
                onSubmit={[Function]}
                title="John"
              />
              <RoleCard
                deleteRole={[Function]}
                match={
                  Object {
                    "url": "",
                  }
                }
                onChange={[Function]}
                onSubmit={[Function]}
                title="move"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
